fixes #10937 - Exposes VM interfaces through API

hufman opened the pull request:
`Host.vm_compute_attributes` calls out to Fog's `Server.attributes`, which omits the network `interfaces` and storage `controllers` data. This will check for those values and add them in to the response.

ares commented:
[test]

domcleal commented on test/unit/compute_resource_test.rb:
```diff
@@ -170,4 +170,43 @@ def setup
+    ComputeResource.any_instance.expects(:find_vm_by_uuid).returns(fake_vm)
+    cr = FactoryGirl.build(:compute_resource)
+    returned_vm = cr.vm_compute_attributes_for('test')
+    assert returned_vm[:cpus] == 2
```
Style nitpick, prefer `assert_equal 2, returned_vm[:cpus]` as it provides better error messages when the assertion fails (showing the actual value).

hufman replied:
Can do!

domcleal commented on test/unit/compute_resource_test.rb:
```diff
@@ -170,4 +170,43 @@ def setup
     refute as_admin { cr.send(:associate_by, 'mac', '11:22:33:44:55:1a') }.readonly?
   end
+
+  test "vm_compute_attributes_for works" do
```
"vm_compute_attributes_for returns attributes hash" ?

hufman replied:
Can do!

domcleal commented on test/unit/compute_resource_test.rb:
```diff
@@ -170,4 +170,43 @@ def setup
+    assert returned_vm[:cpus] == 2
+    refute returned_vm.has_key? :volumes
+    assert returned_vm[:interfaces].length == 1
+    assert returned_vm[:interfaces][0][:mac] == "00:50:56:a5:1f:60"
```
You could probably replace all of this with:

```
assert_equal {:cpus => 2, :interfaces => [{:mac => '00:55:56:a5:1f:60'}]}, returned_vm
```

hufman replied:
Can do!

domcleal commented:
Hm, I just realised this is very similar to #2380 which makes almost exactly the same change to vm_compute_attributes_for.

The only thing it doesn't do is add interfaces, which might be because they're not desired for cloning?  It's also been tested on a wider range of compute resources and has a couple of extra fixes to handle per-CR oddities.  The main difference will be calling the hash "volumes" versus "volumes_attributes".

@tstrachota should we add "interfaces" into the other PR or rebase this one on top it?  Will there be any issues with cloning if we add interfaces in, and should we exclude them?

hufman commented:
This builds on top of #2440 to expose the hosts's volumes and interfaces information through the REST API, and seemed like the smallest change to enable it in my environment. If there is a cleaner method to accomplish it, I'd be happy to do that instead!

domcleal commented:
Sure, it's not that it's unclean, it's just that we have the same change going on with slight differences in two places :)  The other one's a lot more thorough with handling volumes from different compute resource types, but we might need to address the naming and inclusion of interfaces.

I'll try and get some attention on the other PR to figure out what we should do.

tstrachota commented:
@domcleal @hufman it indeed does more or less the same. I commented about it in https://github.com/theforeman/foreman/pull/2380#issuecomment-117562895

domcleal commented:
We've merged #2380 now, sorry for the delay.  Would you mind rebasing this PR?  I think you may want to rename volumes_attributes to volumes, possibly in the controller?

hufman commented:
Ok, I've updated my PR to match that style. I like the better test layout, so I copied it :)

domcleal commented:
Thanks - did you get the chance to read @tstrachota's comment over at https://github.com/theforeman/foreman/pull/2380#issuecomment-117562895?

He's suggesting that we look at either putting the interfaces compute attrs into the response under the interfaces APIs or do it outside of the vm_attributes_for method.  Alternatively we could leave it where it is in this PR and exclude interfaces data from being used during cloning (for now anyway).

[test]

Commit:
Fixes #10937 - Exposes VM interface attributes

This allows a machine's NIC compute_attributes
to be accessible through the REST API

hufman commented:
Ok, I agree with wanting to keep the API output consistent with the create POST bodies. I found the NIC's compute_attributes and exposed them through the hosts/{id}/interfaces/ responses. As a bonus, it also shows up in hosts/{id} in the interfaces section of the response, is that ok?

mmoll commented:
[test]

hufman commented:
Hmmm, I'm not sure how to add the stub properly. It seems that I, and the person who added the other assert in the "show an interface" test, assumed the test would return the `valid_attrs` from the beginning of the file. If I change the test to use assert_equal instead of just assert [boolean], it fails.

I'm unfamiliar with Rails unit testing, can someone give me a quick hint?

domcleal commented on test/functional/api/v2/interfaces_controller_test.rb:
```diff
@@ -24,6 +24,7 @@ def setup
     show_response = ActiveSupport::JSON.decode(@response.body)
     assert !show_response.empty?
     assert "bmc", show_response["type"]
```
This is actually incorrect, it should be `assert_equal "bmc", show_response["type"]` as it's just testing that "bmc" is trueish right now.

domcleal commented:
I can't get this branch to work, doesn't it need to call vm_compute_attributes or similar to populate compute_attributes?

> I found the NIC's compute_attributes and exposed them through the hosts/{id}/interfaces/ responses. As a bonus, it also shows up in hosts/{id} in the interfaces section of the response, is that ok?

My only concern is that if it has to go to the compute resource to get them, then it could be rather slow (esp VMware), so a separate route like the one you added for VM level attributes may be good.  Or add it to the show.rabl in interfaces, so it's only shown for an individual interface?

> Hmmm, I'm not sure how to add the stub properly. It seems that I, and the person who added the other assert in the "show an interface" test, assumed the test would return the valid_attrs from the beginning of the file. If I change the test to use assert_equal instead of just assert [boolean], it fails.

I don't think it's testing against valid_attrs, just that it has a JSON response.  You would need to add a stub on the NIC before the request, something like:

```
@nic.expects(:compute_attributes).returns({'test' => 'yes'})
```

tbrisker commented:
@hufman,
Thank you very much for your efforts!
This PR has been open for a very long time with no activity.
Please indicate if you plan to continue working on it or not; if not, feel free to close it.
If you do not respond either way within a couple of weeks, I will assume this PR is abandoned and will go ahead and close it for you.

tbrisker commented:
Dear @hufman,
It has been 20 days since my last comment with no replies. 
I am assuming this pull request has been abandoned and closing it.
If you wish to continue working on this in the future, feel free to reopen.
On behalf of the entire Foreman team, I wish to thank you for your efforts here and hope to see more contributions from you in the future.
